File: tell-dont-ask-kata-kotlin/src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCase.kt
package main.java.it.gabrieletondi.telldontaskkata.useCase

import main.java.it.gabrieletondi.telldontaskkata.domain.OrderStatus
import main.java.it.gabrieletondi.telldontaskkata.repository.OrderRepository
import main.java.it.gabrieletondi.telldontaskkata.service.ShipmentService


class OrderShipmentUseCase(private val orderRepository: OrderRepository, private val shipmentService: ShipmentService) {

    fun run(request: OrderShipmentRequest) {
        val order = orderRepository.getById(request.orderId)

        if (order.status!! == OrderStatus.CREATED || order.status!! == OrderStatus.REJECTED) {
            throw OrderCannotBeShippedException()
        }

        if (order.status!! == OrderStatus.SHIPPED) {
            throw OrderCannotBeShippedTwiceException()
        }

        shipmentService.ship(order)

        order.status = OrderStatus.SHIPPED
        orderRepository.save(order)
    }
}

File: tell-dont-ask-kata-csharp/TellDontAskKata/useCase/OrderShipmentUseCase.cs
public class OrderShipmentUseCase
{
    private readonly OrderRepository orderRepository;
    private readonly ShipmentService shipmentService;

    public OrderShipmentUseCase(OrderRepository orderRepository, ShipmentService shipmentService)
    {
        this.orderRepository = orderRepository;
        this.shipmentService = shipmentService;
    }

    public void Run(OrderShipmentRequest request)
    {
        Order order = orderRepository.GetById(request.GetOrderId());

        if (order.GetStatus().Equals(OrderStatus.CREATED) || order.GetStatus().Equals(OrderStatus.REJECTED))
        {
            throw new OrderCannotBeShippedException();
        }

        if (order.GetStatus().Equals(OrderStatus.SHIPPED))
        {
            throw new OrderCannotBeShippedTwiceException();
        }

        shipmentService.Ship(order);

        order.SetStatus(OrderStatus.SHIPPED);
        orderRepository.Save(order);
    }
}